File: exam-slammr-webapp/src/main/java/uk/co/claritysoftware/exam/slammr/webapp/security/springsocial/DynamoDBConnectionRepository.java
package uk.co.claritysoftware.exam.slammr.webapp.security.springsocial;

import java.util.List;
import java.util.stream.Collectors;
import org.springframework.dao.DuplicateKeyException;
import org.springframework.security.crypto.encrypt.TextEncryptor;
import org.springframework.social.connect.Connection;
import org.springframework.social.connect.ConnectionData;
import org.springframework.social.connect.ConnectionFactory;
import org.springframework.social.connect.ConnectionFactoryLocator;
import org.springframework.social.connect.ConnectionKey;
import org.springframework.social.connect.ConnectionRepository;
import org.springframework.social.connect.DuplicateConnectionException;
import org.springframework.util.MultiValueMap;
import uk.co.claritysoftware.exam.slammr.webapp.persistence.dynamodb.item.UserConnectionItem;
import uk.co.claritysoftware.exam.slammr.webapp.persistence.dynamodb.repository.DynamoDbUserConnectionItemRepository;

/**
 * TODO - this should use a service rather than the repo directly
 * Implementation of {@link ConnectionRepository} that uses AWS DynamoDB as its persistence store
 */
public class DynamoDBConnectionRepository implements ConnectionRepository {

	private final String userId;

	private final DynamoDbUserConnectionItemRepository userConnectionRepository;

	private final ConnectionFactoryLocator connectionFactoryLocator;

	private final TextEncryptor textEncryptor;

	public DynamoDBConnectionRepository(String userId, DynamoDbUserConnectionItemRepository userConnectionRepository, ConnectionFactoryLocator connectionFactoryLocator, TextEncryptor textEncryptor) {
		this.userId = userId;
		this.userConnectionRepository = userConnectionRepository;
		this.connectionFactoryLocator = connectionFactoryLocator;
		this.textEncryptor = textEncryptor;
	}

	@Override
	public MultiValueMap<String, Connection<?>> findAllConnections() {
		throw new UnsupportedOperationException();
	}

	@Override
	public List<Connection<?>> findConnections(String providerId) {
		List<UserConnectionItem> userConnectionItems = userConnectionRepository.findAllByProviderId(providerId);
		return userConnectionItems.stream()
				.map(this::createConnection)
				.collect(Collectors.toList());
	}

	@Override
	public <A> List<Connection<A>> findConnections(Class<A> apiType) {
		throw new UnsupportedOperationException();
	}

	@Override
	public MultiValueMap<String, Connection<?>> findConnectionsToUsers(MultiValueMap<String, String> providerUserIds) {
		throw new UnsupportedOperationException();
	}

	@Override
	public Connection<?> getConnection(ConnectionKey connectionKey) {
		throw new UnsupportedOperationException();
	}

	@Override
	public <A> Connection<A> getConnection(Class<A> apiType, String providerUserId) {
		throw new UnsupportedOperationException();
	}

	@Override
	public <A> Connection<A> getPrimaryConnection(Class<A> apiType) {
		throw new UnsupportedOperationException();
	}

	@Override
	public <A> Connection<A> findPrimaryConnection(Class<A> apiType) {
		throw new UnsupportedOperationException();
	}

	@Override
	public void addConnection(Connection<?> connection) {
		try {
			ConnectionData data = connection.createData();
			int rank = 1; // TODO - jdbcTemplate.queryForObject("select coalesce(max(rank) + 1, 1) as rank from " + tablePrefix + "UserConnection where userId = ? and providerId = ?", new Object[]{ userId, data.getProviderId() }, Integer.class);

			UserConnectionItem userConnectionItem = UserConnectionItem.builder()
					.userId(userId)
					.providerId(data.getProviderId())
					.providerUserId(data.getProviderUserId())
					.rank(rank)
					.displayName(data.getDisplayName())
					.profileUrl(data.getProfileUrl())
					.imageUrl(data.getImageUrl())
					.accessToken(encrypt(data.getAccessToken()))
					.secret(encrypt(data.getSecret()))
					.refreshToken(encrypt(data.getRefreshToken()))
					.expireTime(data.getExpireTime())
					.build();
			userConnectionRepository.save(userConnectionItem);
		} catch (DuplicateKeyException e) {
			throw new DuplicateConnectionException(connection.getKey());
		}

	}

	@Override
	public void updateConnection(Connection<?> connection) {
		ConnectionData data = connection.createData();
		UserConnectionItem userConnectionItem = userConnectionRepository.findByUserIdAndProviderIdAndProviderUserId(userId, data.getProviderId(), data.getProviderUserId());
		UserConnectionItem updatedUserConnectionItem = userConnectionItem.copy()
				.displayName(data.getDisplayName())
				.profileUrl(data.getProfileUrl())
				.imageUrl(data.getImageUrl())
				.accessToken(encrypt(data.getAccessToken()))
				.secret(encrypt(data.getSecret()))
				.refreshToken(encrypt(data.getRefreshToken()))
				.expireTime(data.getExpireTime())
				.build();
		userConnectionRepository.save(updatedUserConnectionItem);
	}

	@Override
	public void removeConnections(String providerId) {
		throw new UnsupportedOperationException();
	}

	@Override
	public void removeConnection(ConnectionKey connectionKey) {
		throw new UnsupportedOperationException();
	}

	private String encrypt(String text) {
		return text != null ? textEncryptor.encrypt(text) : text;
	}

	private String decrypt(String encryptedText) {
		return encryptedText != null ? textEncryptor.decrypt(encryptedText) : encryptedText;
	}

	private Long expireTime(long expireTime) {
		return expireTime == 0 ? null : expireTime;
	}

	private Connection<?> createConnection(UserConnectionItem userConnectionItem) {
		ConnectionData connectionData = new ConnectionData(
				userConnectionItem.getProviderId(), userConnectionItem.getProviderUserId(), userConnectionItem.getDisplayName(),
				userConnectionItem.getProfileUrl(), userConnectionItem.getImageUrl(), decrypt(userConnectionItem.getAccessToken()),
				decrypt(userConnectionItem.getSecret()), decrypt(userConnectionItem.getRefreshToken()), expireTime(userConnectionItem.getExpireTime()));
		ConnectionFactory<?> connectionFactory = connectionFactoryLocator.getConnectionFactory(connectionData.getProviderId());
		return connectionFactory.createConnection(connectionData);
	}

}

File: exam-slammr-webapp/src/main/kotlin/uk/co/claritysoftware/exam/slammr/webapp/security/springsocial/DynamoDBConnectionRepository.kt
package uk.co.claritysoftware.exam.slammr.webapp.security.springsocial

import org.springframework.dao.DuplicateKeyException
import org.springframework.security.crypto.encrypt.TextEncryptor
import org.springframework.social.connect.*
import org.springframework.util.MultiValueMap
import uk.co.claritysoftware.exam.slammr.webapp.persistence.dynamodb.item.user.UserConnectionItem
import uk.co.claritysoftware.exam.slammr.webapp.persistence.dynamodb.repository.DynamoDbUserConnectionItemRepository
import java.time.Instant
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.util.stream.Collectors

/**
 * TODO - this should use a service rather than the repo directly
 * Implementation of [ConnectionRepository] that uses AWS DynamoDB as its persistence store
 */
class DynamoDBConnectionRepository(val userId: String, val userConnectionRepository: DynamoDbUserConnectionItemRepository, val connectionFactoryLocator: ConnectionFactoryLocator, val textEncryptor: TextEncryptor) : ConnectionRepository {

	override fun findAllConnections(): MultiValueMap<String, Connection<*>> {
		throw UnsupportedOperationException()
	}

	override fun findConnections(providerId: String): List<Connection<*>> {
		val userConnectionItems = userConnectionRepository.findAllByProviderId(providerId)
		return userConnectionItems.stream()
				.map { userConnectionItem -> this.createConnection(userConnectionItem) }
				.collect(Collectors.toList())
	}

	override fun <A> findConnections(apiType: Class<A>): List<Connection<A>> {
		throw UnsupportedOperationException()
	}

	override fun findConnectionsToUsers(providerUserIds: MultiValueMap<String, String>): MultiValueMap<String, Connection<*>> {
		throw UnsupportedOperationException()
	}

	override fun getConnection(connectionKey: ConnectionKey): Connection<*> {
		throw UnsupportedOperationException()
	}

	override fun <A> getConnection(apiType: Class<A>, providerUserId: String): Connection<A> {
		throw UnsupportedOperationException()
	}

	override fun <A> getPrimaryConnection(apiType: Class<A>): Connection<A> {
		throw UnsupportedOperationException()
	}

	override fun <A> findPrimaryConnection(apiType: Class<A>): Connection<A> {
		throw UnsupportedOperationException()
	}

	override fun addConnection(connection: Connection<*>) {
		try {
			val data = connection.createData()
			val rank = 1 // TODO - jdbcTemplate.queryForObject("select coalesce(max(rank) + 1, 1) as rank from " + tablePrefix + "UserConnection where userId = ? and providerId = ?", new Object[]{ userId, data.getProviderId() }, Integer.class);

			val userConnectionItem = UserConnectionItem(
					userId = userId,
					providerId = data.providerId,
					providerUserId = data.providerUserId,
					rank = rank,
					displayName = data.displayName,
					profileUrl = data.profileUrl,
					imageUrl = data.imageUrl,
					accessToken = encrypt(data.accessToken),
					secret = encrypt(data.secret),
					refreshToken = encrypt(data.refreshToken),
					expireTime = toZonedDateTime(data.expireTime))
			userConnectionRepository.save(userConnectionItem)
		} catch (e: DuplicateKeyException) {
			throw DuplicateConnectionException(connection.key)
		}

	}

	override fun updateConnection(connection: Connection<*>) {
		val data = connection.createData()
		val userConnectionItem = userConnectionRepository.findByUserIdAndProviderIdAndProviderUserId(userId, data.providerId, data.providerUserId)
		val updatedUserConnectionItem = UserConnectionItem(
				userId = userConnectionItem.userId,
				providerId = userConnectionItem.providerId,
				providerUserId = userConnectionItem.providerUserId,
				rank = userConnectionItem.rank,
				displayName = data.displayName,
				profileUrl = data.profileUrl,
				imageUrl = data.imageUrl,
				accessToken = encrypt(data.accessToken),
				secret = encrypt(data.secret),
				refreshToken = encrypt(data.refreshToken),
				expireTime = toZonedDateTime(data.expireTime))
		userConnectionRepository.save(updatedUserConnectionItem)
	}

	override fun removeConnections(providerId: String) {
		throw UnsupportedOperationException()
	}

	override fun removeConnection(connectionKey: ConnectionKey) {
		throw UnsupportedOperationException()
	}

	private fun encrypt(text: String?) = text?.let { textEncryptor.encrypt(it) }

	private fun decrypt(encryptedText: String?) = encryptedText?.let { textEncryptor.decrypt(it) }

	private fun toZonedDateTime(millisSinceEpoch: Long?): ZonedDateTime {
		var millisSinceEpoch = millisSinceEpoch
		if (millisSinceEpoch == null) {
			millisSinceEpoch = 0L
		}
		return ZonedDateTime.ofInstant(Instant.ofEpochMilli(millisSinceEpoch), ZoneOffset.UTC)
	}

	private fun toMillis(expireTime: ZonedDateTime?): Long? {
		return expireTime?.toInstant()?.toEpochMilli()
	}

	private fun createConnection(userConnectionItem: UserConnectionItem): Connection<*> {
		val connectionData = ConnectionData(
				userConnectionItem.providerId,
				userConnectionItem.providerUserId,
				userConnectionItem.displayName,
				userConnectionItem.profileUrl,
				userConnectionItem.imageUrl,
				decrypt(userConnectionItem.accessToken),
				decrypt(userConnectionItem.secret),
				decrypt(userConnectionItem.refreshToken),
				toMillis(userConnectionItem.expireTime))
		val connectionFactory = connectionFactoryLocator.getConnectionFactory(connectionData.providerId)
		return connectionFactory.createConnection(connectionData)
	}

}